Make the currently active menu accessible to events, addressing issue #88

vkichline opened the pull request:
This PR adds a static member to the root class: `M5ez::currentMenu`. This makes the currently active menu accessible to event code.
The change has been tested here and by **UT2UH** (originator of issue #88) and was found to adequately address that issue, although it does depart from the original suggestion.

**Design**

* The change is additive and does not modify any existing behavior.  
* `static ezMenu* currentMenu` is added to the class `M5ez`  
* `M5ez::currentMenu` is set on entry to `runOnce()` whenever a menu is rendered/handled.  
* It is cleared when `runOnce()` returns zero, indicating termination.  
* For additional safety, a destructor is added to `ezMenu`, which also clears `M5ez::currentMenu` if and only if it matches `ezMenu`'s `self`. This is safe, it's cheap, and extra safety when caching is a good policy.

**Benefits**

A pointer to the current menu is generally available, as a member of the global `ez`. This makes it possible for ezEvent code to modify the current menu with asynchronous data, addressing the concern in issue #88.  
In addition, it lays the groundwork for developing a non-blocking menu system, where one might call `ez.update()` in the Arduino `loop()` function, which would address other issues.

**Side Effects**

This approach means that if a menu's `simpleFunction` or `advancedFunction` calls code which renders a screen that does not display a menu at all, `M5ez::currentMenu` would still return a pointer to the (hidden) menu. I believe that this may be more of a benefit than a flaw, but it should be documented. If a `simpleFunction` or `advancedFunction` called code which rendered a sub-menu, `M5ez::currentMenu` would accurately track the change as soon as `runOnce()` is called, rendering the menu for the first time.

**Limitations**

There's no way for the event code to identify the current menu; no identifier is available from the public interface. Should a public `String ezMenu::getHdr()` be added to facilitate identification?

**Concerns**

Are there additional safety checks that could be made if we used a public accessor function for a private member variable instead? I believe it would be possible to write code that would leave a dangling `M5ez::currentMenu`, but haven't found a way. I believe that such code is easy to avoid and would likely be buggy in other respects. Suggested approaches are welcome; I will try them.

**Open Issues**

* Make a call on adding `ezMenu::getHdr()` in this PR.
* The scenario described in Side Effects hasn't been adequately tested. I'll do that.  
* I will augment this PR with documentation changes once this approach is approved or improved.

ropg commented:
Strongly approve. And excellent write-up. Certainly more thoroughly thought about than anything I have ever added to the code ... 👍 

(And it's clever doing it this way instead of passing things around....)

vkichline commented:
I have made a few changes for additional safety:

* Replaced `public ezMenu*M5ez::currentMenu` with `private ezMenu* M5ez::_currentMenu`
* Added `public ezMenu* M5ez::getCurrentMenu()`. This will permit adding safety fixes if they are discovered without changing the public interface.
* Added `String ezMenu::getTitle()` to aid in identifying the current menu. Avoided `getHeader` which might lead one to expect an ezHeader result.

I tested the scenario described above in Side Effects and there were no issues detected.

I'll complete the documentation changes soon. I believe that this information will be more helpful under **Events** than under **Menus**.
